src/lib/foreign-field.ts

```diff
@@ -0,0 +1,496 @@
+import { Snarky } from '../snarky.js';
+import { mod, inverse, Fp } from '../bindings/crypto/finite_field.js';
+import { Tuple } from '../bindings/lib/binable.js';
+import {
+  Field,
+  FieldConst,
+  FieldVar,
+  checkBitLength,
+  withMessage,
+} from './field.js';
+import { Provable } from './provable.js';
+import { Bool } from './bool.js';
+
+// external API
+export { createForeignField, ForeignField };
+
+// internal API
+export { ForeignFieldVar, ForeignFieldConst, limbBits };
+
+const limbBits = 88n;
+
+type MlForeignField<F> = [_: 0, x0: F, x1: F, x2: F];
+type ForeignFieldVar = MlForeignField<FieldVar>;
+type ForeignFieldConst = MlForeignField<FieldConst>;
+type ForeignField = InstanceType<ReturnType<typeof createForeignField>>;
+
+/**
+ * Create a class representing a prime order finite field, which is different from the native {@link Field}.
+ *
+ * ```ts
+ * const SmallField = createForeignField(17n); // the finite field F_17
+ * ```
+ *
+ * `createForeignField(p)` takes the prime modulus `p` of the finite field as input, as a bigint.
+ * We support prime moduli up to a size of 259 bits.
+ *
+ * The returned {@link ForeignField} class supports arithmetic modulo `p` (addition and multiplication),
+ * as well as helper methods like `assertEquals()` and `equals()`.
+ *
+ * _Advanced usage:_
+ *
+ * Internally, a foreign field element is represented as three native field elements, each of which
+ * represents a limb of 88 bits. Therefore, being a valid foreign field element means that all 3 limbs
+ * fit in 88 bits, and the foreign field element altogether is smaller than the modulus p.
+ * With default parameters, new `ForeignField` elements introduced in provable code are automatically
+ * constrained to be valid on creation.
+ *
+ * However, optimized code may want to forgo these automatic checks because in some
+ * situations they are redundant. Skipping automatic validity checks can be done
+ * by passing the `unsafe: true` flag:
+ *
+ * ```ts
+ * class UnsafeField extends createForeignField(17n, { unsafe: true }) {}
+ * ```
+ *
+ * You then often need to manually add validity checks:
+ * ```ts
+ * let x: UnsafeField;
+ * x.assertValidElement(); // prove that x is a valid foreign field element
+ * ```
+ *
+ * @param modulus the modulus of the finite field you are instantiating
+ * @param options
+ * - `unsafe: boolean` determines whether `ForeignField` elements are constrained to be valid on creation.
+ */
+function createForeignField(modulus: bigint, { unsafe = false } = {}) {
+  const p = modulus;
+  const pMl = ForeignFieldConst.fromBigint(p);
+
+  if (p <= 0) {
+    throw Error(`ForeignField: expected modulus to be positive, got ${p}`);
+  }
+  if (p > foreignFieldMax) {
```

Comment:
Thanks! I actually read the RFC and wanted to use the exact number, but sadly I couldn't compute it exactly with JS floats, and I also didn't want to hard-code a number when everything else is derived from simple constants.
So I looked at the difference between 2^259 and the exact max modulus and figured that they are so close that probably no one will ever use a prime modulus larger than 259 bits.

In fact, this is how `926336713898529563388567880069503262826888842373627227613104999999999999999607` looks in hex:
`80000000000000000000000000000000224698fc094cf91b9908b7c12f823fe77`
you can see how close it is to 2^259 which is `80000...000`

Reply:
Also, documenting "259 bits" felt simpler / easier to to use than some random large number

Reply:
> you can see how close it is to 2^259 which is 80000...000

Yeah exactly, that's why I said 259.5 bits, because that number is more or less sqrt of 2^519

I am fine with restricting it further, but it is good to know for us that actually we do have some margin in case we want to use it in another scenario where 259.5 bits are needed.

Reply:
Cool! The max modulus is not that close to the sqrt of 2^519 though. It's much closer to the sqrt of 2^518, which is 2^259, because the pasta moduli are very close to 2^254 (not 2^255)

Reply:
Well, just for curiosity, the exact number is computed performing `sqrt(2^264*pallas)` which is about `sqrt(2^519)`, that's why I say 259.5 bits. Then that modulus is the maximum prime number that still satisfies that.

Reply:
`pallas` is about 2^254